components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/JDBCSubscriberDAOImpl.java

```diff
@@ -0,0 +1,372 @@
+            throw new EntitlementException(ERROR_SUBSCRIBER_ID_NULL);
+        }
+
+        if (isSubscriberExists(subscriberId)) {
+            throw new EntitlementException("Subscriber ID already exists");
+        }
+
+        PublisherPropertyDTO[] propertyDTOs = holder.getPropertyDTOs();
+        insertSubscriber(subscriberId, holder, propertyDTOs);
+    }
+
+    /**
+     * Updates a subscriber.
+     *
+     * @param holder publisher data holder.
+     * @throws EntitlementException If an error occurs.
+     */
+    @Override
+    public void updateSubscriber(PublisherDataHolder holder) throws EntitlementException {
+
+        String subscriberId = EntitlementUtil.resolveSubscriberId(holder);
+        if (subscriberId == null) {
+            throw new EntitlementException(ERROR_SUBSCRIBER_ID_NULL);
+        }
+
+        PublisherDataHolder oldHolder;
+        if (isSubscriberExists(subscriberId)) {
+            // TODO: revisit in hybrid impl
+            oldHolder = getSubscriber(subscriberId, false);
+            populateProperties(holder, oldHolder);
+        } else {
+            throw new EntitlementException("Subscriber ID does not exist; update cannot be done");
+        }
+
+        PublisherPropertyDTO[] propertyDTOs = holder.getPropertyDTOs();
+        updateSubscriber(subscriberId, holder, propertyDTOs, oldHolder);
+    }
+
+    /**
+     * Removes the subscriber of the given subscriber ID.
+     *
+     * @param subscriberId subscriber ID.
+     * @throws EntitlementException If an error occurs.
+     */
+    @Override
+    public void removeSubscriber(String subscriberId) throws EntitlementException {
+
+        int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+
+        if (StringUtils.isBlank(subscriberId)) {
+            throw new EntitlementException(ERROR_SUBSCRIBER_ID_NULL);
+        }
+
+        if (EntitlementConstants.PDP_SUBSCRIBER_ID.equals(subscriberId.trim())) {
+            throw new EntitlementException("Cannot delete PDP publisher");
+        }
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false)) {
+            try (NamedPreparedStatement preparedStmt = new NamedPreparedStatement(connection, DELETE_SUBSCRIBER_SQL)) {
+                preparedStmt.setString(SUBSCRIBER_ID, subscriberId);
+                preparedStmt.setInt(TENANT_ID, tenantId);
+                preparedStmt.executeUpdate();
+            }
+        } catch (SQLException e) {
+            throw new EntitlementException("Error while deleting subscriber details", e);
+        }
+    }
+
+    /**
+     * Checks whether a subscriber exists.
+     *
+     * @param subscriberId subscriber ID.
+     * @return whether the subscriber exists or not.
+     * @throws EntitlementException If an error occurs.
+     */
+    public boolean isSubscriberExists(String subscriberId) throws EntitlementException {
+
+        int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
```

Comment:
Shall we pass `tenantId` as a param to keep consistency of other pure DAO methods?